cmd/agent/api/server.go

```diff
@@ -69,124 +68,64 @@ func StartServer(
-		workloadmetaServer: workloadmetaServer.NewServer(wmeta),
-		dogstatsdServer:    dogstatsdServer,
-		capture:            capture,
-	})
-
-	dcreds := credentials.NewTLS(&tls.Config{
-		ServerName: tlsAddr,
-		RootCAs:    tlsCertPool,
-	})
-	dopts := []grpc.DialOption{grpc.WithTransportCredentials(dcreds)}
-
-	// starting grpc gateway
-	ctx := context.Background()
-	gwmux := runtime.NewServeMux()
-	err = pb.RegisterAgentHandlerFromEndpoint(
-		ctx, gwmux, tlsAddr, dopts)
+	tlsKeyPair, tlsCertPool, err := initializeTLS(extraHosts...)
 	if err != nil {
-		return fmt.Errorf("error registering agent handler from endpoint %s: %v", tlsAddr, err)
+		return fmt.Errorf("unable to initialize TLS: %v", err)
 	}
 
-	err = pb.RegisterAgentSecureHandlerFromEndpoint(
-		ctx, gwmux, tlsAddr, dopts)
-	if err != nil {
-		return fmt.Errorf("error registering agent secure handler from endpoint %s: %v", tlsAddr, err)
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{*tlsKeyPair},
+		NextProtos:   []string{"h2"},
+		MinVersion:   tls.VersionTLS12,
 	}
 
-	// Setup multiplexer
-	// create the REST HTTP router
-	agentMux := gorilla.NewRouter()
-	checkMux := gorilla.NewRouter()
-	// Validate token for every request
-	agentMux.Use(validateToken)
-	checkMux.Use(validateToken)
-
-	mux := http.NewServeMux()
-	mux.Handle(
-		"/agent/",
-		http.StripPrefix("/agent",
-			agent.SetupHandlers(
-				agentMux,
-				flare,
-				dogstatsdServer,
-				serverDebug,
-				wmeta,
-				logsAgent,
-				senderManager,
-				hostMetadata,
-				invAgent,
-				demux,
-				invHost,
-				secretResolver,
-			)))
-	mux.Handle("/check/", http.StripPrefix("/check", check.SetupHandlers(checkMux)))
-	mux.Handle("/", gwmux)
-
-	// Use a stack depth of 4 on top of the default one to get a relevant filename in the stdlib
-	logWriter, _ := config.NewLogWriter(5, seelog.ErrorLvl)
-
-	srv := grpcutil.NewMuxedGRPCServer(
-		tlsAddr,
-		&tls.Config{
-			Certificates: []tls.Certificate{*tlsKeyPair},
-			NextProtos:   []string{"h2"},
-			MinVersion:   tls.VersionTLS12,
-		},
-		s,
-		grpcutil.TimeoutHandlerFunc(mux, time.Duration(config.Datadog.GetInt64("server_timeout"))*time.Second),
-	)
+	if err := util.CreateAndSetAuthToken(); err != nil {
+		return err
+	}
 
-	srv.ErrorLog = stdLog.New(logWriter, "Error from the agent http API server: ", 0) // log errors to seelog
+	if err := startCMDServer(
+		apiAddr, tlsConfig, tlsCertPool,
+		configService, flare, dogstatsdServer,
+		capture, serverDebug, wmeta, logsAgent,
+		senderManager, hostMetadata, invAgent,
+		demux, invHost, secretResolver,
+	); err != nil {
+		return err
+	}
 
-	tlsListener := tls.NewListener(listener, srv.TLSConfig)
+	if apiConfigEnabled {
+		if err := startIPCServer(apiConfigHostPort, tlsConfig); err != nil {
+			StopServer()
+			return err
+		}
+	}
 
-	go srv.Serve(tlsListener) //nolint:errcheck
 	return nil
 }
 
 // StopServer closes the connection and the server
 // stops listening to new commands.
 func StopServer() {
```

Comment:
Similar note that we may want to rename this since we're starting multiple servers now

Reply:
done